jsonapi_review.go

```diff
@@ -0,0 +1,547 @@
+    included, err := marshalIncluded(mi)
+    if err != nil {
+      return nil, err
+    }
+
+    for _, inc := range included {
+      for _, one := range inc {
+        doc.Included = append(doc.Included, one)
+      }
+    }
+  }
+
+  mm, ok := payload.(MarshalMeta)
+  if ok {
+    meta, err := marshalMeta(mm)
+    if err != nil {
+      return nil, err
+    }
+
+    doc.Meta = meta
+  }
+
+  return doc, nil
+}
+
+func marshalDocumentSlice(payload interface{}) (*Document, error) {
+  var doc *Document
+
+  if errorObjects, ok := payload.([]*ErrorObject); ok {
+    doc = &Document{
+      Errors: errorObjects,
+    }
+  } else {
+    doc = &Document{
+      Data: &documentData{
+        Many: []*ResourceObject{},
+      },
+    }
+
+    many, err := marshalResourceObjects(payload)
+    if err != nil {
+      return nil, err
+    }
+
+    doc.Data.Many = many
+
+    included := make(map[string]map[string]*ResourceObject)
+
+    if mi, ok := payload.(MarshalIncluded); ok {
+      inc, err := marshalIncluded(mi)
+      if err != nil {
+        return nil, err
+      }
+
+      for typ, value := range inc {
+        if _, ok := included[typ]; !ok {
+          included[typ] = make(map[string]*ResourceObject)
+        }
+
+        for id, one := range value {
+          if _, ok := included[typ][id]; !ok {
+            included[typ][id] = one
+          }
+        }
+      }
+    } else {
+      value := reflect.ValueOf(payload)
+
+      for i := 0; i < value.Len(); i++ {
+        if mi, ok := value.Index(i).Interface().(MarshalIncluded); ok {
+          inc, err := marshalIncluded(mi)
+          if err != nil {
+            return nil, err
+          }
+
+          for typ, value := range inc {
+            if _, ok := included[typ]; !ok {
+              included[typ] = make(map[string]*ResourceObject)
+            }
+
+            for id, one := range value {
+              if _, ok := included[typ][id]; !ok {
+                included[typ][id] = one
+              }
+            }
+          }
+        }
+      }
+    }
+
+    for _, inc := range included {
+      for _, one := range inc {
+        doc.Included = append(doc.Included, one)
+      }
+    }
+
+    if mm, ok := payload.(MarshalMeta); ok {
+    	meta, err := marshalMeta(mm)
+    	if err != nil {
+    		return nil, err
+    	}
+      doc.Meta = meta
+    }
+  }
+  return doc, nil
+}
+
+func marshalResourceObjectIdentifier(mri MarshalResourceIdentifier) ResourceObjectIdentifier {
+  return ResourceObjectIdentifier{ ID: mri.GetID(), Type: mri.GetType() }
+}
+
+func marshalResourceObject(mri MarshalResourceIdentifier) (ResourceObject, error) {
+  one := ResourceObject{
+    ResourceObjectIdentifier: marshalResourceObjectIdentifier(mri),
+  }
+
+  buf := &bytes.Buffer{}
+	enc := json.NewEncoder(buf)
+	enc.SetEscapeHTML(false)
+
+  err := enc.Encode(mri)
+  if err != nil {
+    return one, err
+  }
+
+  if !bytes.Equal(buf.Bytes(), []byte("{}\n")) {
```

Comment:
not sure about this check, at least I would check for size of bytes, like if it's 3 bytes or less (covers missing \n etc)